tests/analysis/impl/dense/gpcca_test.py

```diff
@@ -0,0 +1,681 @@
+
+        np.testing.assert_array_equal(kopt, [3, 3, 3, 3, 2, 2, 7])
+
+    def test_cluster_by_first_col_not_1(self):
+        svecs = np.zeros((4, 3))
+        svecs[0, 0] = 1
+
+        with pytest.raises(
+            ValueError,
+            match="First Schur vector is not constant. This indicates that the Schur vectors are incorrectly sorted. "
+            "Cannot search for a simplex structure in the data",
+        ):
+            _cluster_by_isa(svecs)
+
+    def test_cluster_by_isa_shape_error(self):
+        svecs = np.zeros((3, 4))
+        svecs[:, 1] = 1.0
+
+        with pytest.raises(
+            ValueError,
+            match=r"The Schur vector matrix of shape \(\d+, \d+\) has more columns than rows. You can't get a "
+            r"\d+-dimensional simplex from \d+ data vectors.",
+        ):
+            _cluster_by_isa(svecs)
+
+    def test_cluster_by_isa(
+        self, chi_isa_mu0_n3: np.ndarray, chi_isa_mu100_n3: np.ndarray
+    ):
+        # chi_sa_mu0_n3 has permuted 2nd and 3d columns when compared to the matlab version
+        for m, chi_exp in zip([0, 100], [chi_isa_mu0_n3, chi_isa_mu100_n3]):
+            mu_ = mu(m)
+            P, sd = get_known_input(mu_)
+            X, _, _ = _do_schur(P, sd, m=3)
+            chi, _ = _cluster_by_isa(X[:, :3])
+
+            chi = chi[:, _find_permutation(chi_exp, chi)]
+
+            assert_allclose(chi.T @ chi, chi_exp.T @ chi_exp)
+            assert_allclose(chi, chi_exp)
+
+    def test_use_minChi(self):
+        kmin, kmax = 2, 9
+        kopt = [None] * 7
+
+        for i, m in enumerate([0, 10, 50, 100, 200, 500, 1000]):
+            mu_ = mu(m)
+            P, sd = get_known_input(mu_)
+            g = GPCCA(P, eta=sd)
+            minChi = g.minChi(kmin, kmax)
+
+            kopt[i] = kmax - 1 - np.argmax(np.flipud(minChi[1:-1]))
+
+        np.testing.assert_array_equal(kopt, [3] * 6 + [7])
+
+    def test_gpcca_brandts_sparse_is_densified(self, P: np.ndarray, sd: np.ndarray):
+
+        with pytest.warns(
+            UserWarning,
+            match=r"Sparse implementation is only avaiable for `method='krylov'`, densifying.",
+        ):
+            GPCCA(csr_matrix(P), eta=sd, method="brandts").optimize(3)
+
+    def test_sort_real_schur(self, R_i: np.ndarray):
+        # test_SRSchur_num_t
+        Q = np.eye(4)
+        QQ, RR, ap = sort_real_schur(Q, R_i, z="LM", b=0)
+
+        assert np.all(np.array(ap) <= 1), ap
+
+        EQ = np.true_divide(np.linalg.norm(Q - QQ.T @ QQ, ord=1), eps)
+        assert_allclose(EQ, 1.0, atol=5)
+
+        EA = np.true_divide(np.linalg.norm(R_i - QQ @ RR @ QQ.T, ord=1), eps * np.linalg.norm(R_i, ord=1))
+        assert_allclose(EA, 1.0, atol=5)
+
+        l = eigs(R_i, k=4, which="SM", return_eigenvectors=False)
+        ll = eigs(RR, k=4, which="SM", return_eigenvectors=False)
+        EL = np.true_divide(np.abs(l - ll), eps * np.abs(l))
+        assert_allclose(EL, 1.0, atol=5)
```

Comment:
@msmdev , this one fails and we don't understand what it really does. Can you help us here please?

Reply:
How does it fail exactly? What is the output?